agent/connectionuptime/service.go

```diff
@@ -0,0 +1,156 @@
+// pmm-agent
+// Copyright 2019 Percona LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connectionuptime
+
+import (
+	"math/big"
+	"time"
+)
+
+const channelBufferSize = 10
+
+// Service calculates connection uptime between agent and server
+type Service struct {
+	uptimeSeconds big.Int
```

Comment:
What is the reason to use big.Int instead of a slice of times?
the current implementation in most cases uses more CPU and more memory and at the same time harder to understand. The previous implementation could give us more opportunities to improve debug logic by including a whole array of connection and disconnection logic to logs in the future.

Reply:
>the current implementation in most cases uses more CPU and more memory and at the same time harder to understand

Implementation with bitmap (big.Int) will be very efficient.

Reply:
For the one-day time range, it uses 86400 bits which is 10800 bytes. Isn't it too much?
Looping through this range to get the uptime value also doesn't look efficient.

Reply:
I said slices of times, not booleans.

Reply:
Okay, but slice of `time.Time` even will consume much more memory then slice of booleans, because `time.Time` struct has two variables with size of 8 bytes and one pointer. 

I guess from the perspective of memory consumption bit set will be much better.

what do you think?

Reply:
No, I don't

I already changed the implementation of calculation connection uptime.

Now here will be stored practicallly a ring array and which element of it it means a second. And value means whether it's connected or not. So if specified window time is hour then we should allocate 3600 elements.

Every time when we register a new event, then we set the number of elements which is equal to difference between current timesampt and timestamp of previouse call in seconds.

And for calculation connection up time we just divide the number of elements which makred as `connected` (has 1 value in this bitset) to whole size of bitset (or number of passed time in seconds if it is less then the window time).

Reply:
> Do you expect that your system will connect-disconnect more often than each minute?

That is entirely possible

Reply:
> That is entirely possible, if the requests are frequent enough anyway.

Okay, and it can be handled by this implementation as well.

Reply:
@cezmunsta That sounds like another issue, barely related to this one. which might happen because of too many requests.
Could you please explain how it can bring to a chain of very often successful and unsuccessful connections? Like connected-disconnected-connected-disconnected-connected-disconnected